Amend RFD 0159 to add some details about the WebAuthn flow change

bl-nero opened the pull request:
Amending per @codingllama's request.

bl-nero committed:
Amend RFD 0159 to add some details about the WebAuthn flow change

codingllama commented on rfd/0159-password-status.md:
```diff
@@ -66,6 +68,48 @@ ContactAdmin --> End([End])
 
+We are going to implement the following changes to the WebAuthn flow and APIs:
+
+1. Move the responsibility for verifying the old password to the `ChangePassword` RPC call. This will allow us to decouple the verification from setting the new password in the UI (the old password will not be required when generating the challenge). This verification will be specific to the `ChangePassword` call and will not be a part of the generic `LoginFlow`. The
```
Leftover word?

```suggestion
1. Move the responsibility for verifying the old password to the `ChangePassword` RPC call. This will allow us to decouple the verification from setting the new password in the UI (the old password will not be required when generating the challenge). This verification will be specific to the `ChangePassword` call and will not be a part of the generic `LoginFlow`.
```

codingllama commented on rfd/0159-password-status.md:
````diff
@@ -66,6 +68,48 @@ ContactAdmin --> End([End])
+     string user_verification_requirement = 3; // "required", "discouraged", etc.
+   }
+   ```
+   This will automatically become a parameter of the `CreateAuthenticateChallenge` RPC call through the `ChallengeExtensions` field. To make it possible to specify from the client side, we will accept a corresponding `userVerificationRequirement` parameter to the `/v1/webapi/mfa/authenticatechallenge/password` POST endpoint.
````
Is there a reason for keeping `/v1/webapi/mfa/authenticatechallenge/password` at this point? Could we remove it and use `/v1/webapi/mfa/authenticatechallenge` in its place?

bl-nero replied:
Good question. I felt like it's less risky to allow explicitly setting the verification requirement parameter just in place where we actually need it. If you think it's not adding unnecessary risk, we can remove this thing altogether. (And on the second thought, if we're worried about allowing this for every type of challenge, we can always only allow it for the password change scope.

codingllama replied:
Good points, thanks for looking at it more closely.

Let me give it a try at answering:

1. IsMFARequired can't be used to bypass a check, just to check whether an MFA check is necessary. This is usually targeted at resources. It's like asking "do I need MFA to SSH to this node?" or similar questions. Not really relevant here.

2. I believe setting AllowReuse for PASSWORD_CHANGE would cause a backend error. Even if it passed, the verifier (in our case the ChangePassword RPC) will say that reuse is not allowed, so it wouldn't pass there either.

3. That I didn't know. I suppose we could still do that in the generic RPC by looking at the scope.

In general I prefer removing the /password endpoint and having less API surface. Chances are we'll either need UV in the generic endpoint or some other flag in the /password endpoint, so they will converge, duplicate, and potentially deviate in annoying ways over time.

bl-nero replied:
All right, we will use the generic endpoint. Modified the document appropriately.

codingllama commented on rfd/0159-password-status.md:
```diff
@@ -66,6 +68,48 @@ ContactAdmin --> End([End])
 
+We are going to implement the following changes to the WebAuthn flow and APIs:
+
+1. Move the responsibility for verifying the old password to the `ChangePassword` RPC call. This will allow us to decouple the verification from setting the new password in the UI (the old password will not be required when generating the challenge). This verification will be specific to the `ChangePassword` call and will not be a part of the generic `LoginFlow`. The
```
> will not be a part of the generic `LoginFlow`

I'm not sure your average reader will know what LoginFlow is. Heck, I wouldn't have remember if I wasn't messing with it just this week.

Maybe add a link?

bl-nero replied:
Haha, good point :)

codingllama commented on rfd/0159-password-status.md:
```diff
@@ -66,6 +68,48 @@ ContactAdmin --> End([End])
+P ->> A: ChangePassword
+deactivate P
+activate A
+A ->> A: Verify scope and assertion response
+A ->> A: Verify old password if UV=0
```
Suggestion:

```suggestion
A ->> A: if old password is empty:<br/>UserVerificationRequirement=required
A ->> A: if old password is not empty:<br/>Verify old password
A ->> A: Verify assertion<br/>scope=PASSWORD_CHANGE<br/>UVR as above
```

(I don't know if this renders alright, so feel free to change as you prefer.)

bl-nero replied:
@codingllama I'm sorry, I don't understand what "scope=PASSWORD_CHANGE" and "UVR as above" mean in the last line.

codingllama replied:
When you verify an assertion, like in the ChangePassword RPC, you have to say what are the ChallengeExtensions you expect to see. In our case, we would verify using scope=PASSWORD_CHANGE and UVR="required" (UserVerificationRequirement). The fact that we say what extensions we expect to see is what guarantees the "wrong" challenge won't be accepted.

(UVR is just a shortened "UserVerificationRequirement". Feel free to spell it out.)

codingllama replied:
Btw, this some of this is in RFD 155 - https://github.com/gravitational/teleport/blob/master/rfd/0155-scoped-webauthn-credentials.md.

bl-nero replied:
@codingllama I think I addressed what you wrote, though I made it a bit more clear. Please take a look. (I also included a reference to RFD 155).

codingllama replied:
Looks good!

codingllama reviewed (commented):
Looks good, thanks for the quick follow up.

bl-nero committed:
Fix a leftover word

Co-authored-by: Alan Parra <alan.parra@goteleport.com>

bl-nero committed:
Review

Also: update the hashes in the link to the master branch hash, so that they
never expire after squashing this PR.

bl-nero commented on rfd/0159-password-status.md:
```diff
@@ -41,6 +41,8 @@ Rationale: as there is no reliable way to verify that the user actually has a pa
 
 The user will need to pick upfront whether they want to authenticate with a passwordless or an MFA device, even in case of hardware tokens. This is because in order to perform a WebAuthn authentication ceremony, we need to specify upfront whether we require [user verification](https://w3c.github.io/webauthn/#user-verification) or not using the [`PublicKeyCredentialRequestOptions.userVerification` ...
 
+There are two RPCs that take part in the process of changing password when token authentication is required: the first one that generates the challenge (currently handled by )
```
OH, dear, thank you for noticing it.

bl-nero committed:
Update 0159-password-status.md

Remove a leftover sentence
